### ISSUE TITLE
Doc/question: Adaptive scans with area detectors

### ISSUE THREAD

**Issue by CJ-Wright** [NSLS-II/bluesky#1079]
Reading through the docs for single process [adaptive scans](http://nsls-ii.github.io/bluesky/tutorial.html#adaptive-logic-in-a-plan) everything looks great! However, does the same logic work for area detectors which save files? I'm guessing that the results that the detector responds with will be the filename and/or other information that is passed to the databroker for it to properly register the data. If this is the case what changes are needed to that scan to work with file saving detectors?

**Comment by prjemian** [comment]
Doesn't the adaptive scan need a stream of floats (or ints) on which to base its decisions?  A linear or area detector (some file-saving detector) will need to summarize each image frame into such a stream.  Different strategies could apply such as a series of frames might have a rectangular region of interest while a single frame might generate stream of numbers from a line of pixels drawn across the image.

Bottom line, adaptive scan: does it only work with point detectors?  If we have a n-D detector (n > 0), can we use adaptive scans?

**Comment by CJ-Wright** [comment]
I was mostly using that example as a jumping off point. One could sum over the detector to create a single value or run some other advanced computation.

**Comment by danielballan** [comment]
All of the adaptive scan examples current implemented (including tests, documented examples, usages in production of which I am aware) utilize the communication channel of the RunEngine `send`-ing data back to the plan via `generator.send(reading)`. Indeed, that mechanism does not work on _externally-stored data_*; if the reading in question contains a `datum_id`, the plan won't have enough information to do anything with that. The plan needs to upgrade to a richer communication channel:

```py
def plan():
    cache = []
    def collect(name, doc):
        cache.append((name, doc))
    yield bluesky.plan_stubs.subscribe(collect)
    ...
```

Via the subscription mechanism, the plan now has access to all the Datum and Resource documents emitted by the RunEngine, and it now knows what to do with the uids it finds it receives back from `'read'` Msgs.

*CJ's example is using "area detectors" as a way to say "externally-stored data". The _dimensionality_ itself isn't directly important --- you are allowed to stuff N-D arrays into readings if you want to --- but generally speaking large array are externally stored and utilize the Resource/Datum documents to manage that.

**Comment by teddyrendahl** [comment]
Think this is alluded to in a number of different older issue as well (https://github.com/NSLS-II/bluesky/issues/862, _Somewhat related but caching an output value of a plan not a read_  https://github.com/NSLS-II/bluesky/issues/877)

**Comment by danielballan** [comment]
I think the action items here are:

- [ ] Turn my comment above into documentation of the "recommended" way to do this.
- [ ] For convenience, add a class to bluesky that acts as an in-memory databroker meant to have a short lifetime and to live inside a plan, making it easier to do handler-registry stuff. It will not support search --- usages like `db(operator='Dan')` --- but it will support lookup by recency and uid.

**Comment by CJ-Wright** [comment]
Would it be possible expose something a bit lighter than the databroker? Something callback like, where it takes in documents, fills events, and returns the filled events?

**Comment by danielballan** [comment]
That's the idea. We don't want a databroker dependency, so this would be something databroker-like that supports only a subset of the API and introduces no extra deps.

**Comment by CJ-Wright** [comment]
Ah ok, I guess I didn't think we would need lookup, maybe? Although, I'm ok with the lookup tooling.

**Comment by danielballan** [comment]
I think lookup is not the most important part of this but maybe worth. Multi-run plans aren't uncommon, so some way of grabbing the run you want seems important. I would be fine is uid-based lookup was the only option though.
